src/Type/Php/FilterVarFunctionTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,129 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\BinaryOp\BooleanAnd;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\Analyser\SpecifiedTypes;
+use PHPStan\Analyser\TypeSpecifier;
+use PHPStan\Analyser\TypeSpecifierAwareExtension;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\BooleanType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\FunctionTypeSpecifyingExtension;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use function defined;
+use function in_array;
+use function strtolower;
+use const FILTER_VALIDATE_BOOL;
+use const FILTER_VALIDATE_BOOLEAN;
+use const FILTER_VALIDATE_DOMAIN;
+use const FILTER_VALIDATE_EMAIL;
+use const FILTER_VALIDATE_FLOAT;
+use const FILTER_VALIDATE_INT;
+use const FILTER_VALIDATE_URL;
+
+class FilterVarFunctionTypeSpecifyingExtension implements FunctionTypeSpecifyingExtension, TypeSpecifierAwareExtension
+{
+
+	private TypeSpecifier $typeSpecifier;
+
+	public function isFunctionSupported(
+		FunctionReflection $functionReflection,
+		FuncCall $node,
+		TypeSpecifierContext $context,
+	): bool
+	{
+		return strtolower($functionReflection->getName()) === 'filter_var' && $context->true();
+	}
+
+	public function specifyTypes(
+		FunctionReflection $functionReflection,
+		FuncCall $node,
+		Scope $scope,
+		TypeSpecifierContext $context,
+	): SpecifiedTypes
+	{
+		$args = $node->getArgs();
+		if (count($args) < 2) {
+			return new SpecifiedTypes();
+		}
+
+		$flagsType = $scope->getType($args[1]->value);
+		if (!$flagsType instanceof ConstantIntegerType) {
+			return new SpecifiedTypes();
+		}
+
+		$type = null;
+		if ($flagsType->getValue() === FILTER_VALIDATE_FLOAT) {
+			$type = new FloatType();
+		}
+
+		// note that FILTER_VALIDATE_BOOLEAN tries to be smart, recognizing words like Yes, No, Off, On, both string
+		// and native types of true and false, and is not case-sensitive when validating strings.
+		if (defined('FILTER_VALIDATE_BOOL')) { // requires php 8.0+
+			if (in_array($flagsType->getValue(), [FILTER_VALIDATE_BOOL, FILTER_VALIDATE_BOOLEAN], true)) {
+				$type = TypeCombinator::union(
+					new BooleanType(),
+					new StringType(),
+					new ConstantIntegerType(0),
+					new ConstantIntegerType(1),
+				);
+			}
+		} else {
+			if ($flagsType->getValue() === FILTER_VALIDATE_BOOLEAN) {
+				$type = TypeCombinator::union(
+					new BooleanType(),
+					new StringType(),
+					new ConstantIntegerType(0),
+					new ConstantIntegerType(1),
+				);
+			}
+		}
+
+		if (in_array($flagsType->getValue(), [FILTER_VALIDATE_DOMAIN, FILTER_VALIDATE_URL, FILTER_VALIDATE_EMAIL], true)) {
+			$type = new IntersectionType([
```

Comment:
looks like you missed changing this to use the TypeCombinator still :)

Reply:
in most other places of the codebase we use `new IntersectionType` instead of the `TypeCombinator` for non-empty-string types